editoast: `track_occupancy` for `paced_trains`

hamz2a opened the pull request:
closes https://github.com/OpenRailAssociation/osrd/issues/13260

- [x] add endpoint
- [x] add tests

woshilapin commented on editoast/src/views/timetable/train_schedule.rs:
```diff
@@ -1027,53 +1022,24 @@ fn track_occupancy_on_path_item(
-            .track
-            .to_string(),
-    };
+    let train_schedule_schema: TrainSchedule = train_schedule.clone().into();
```
```suggestion
    let train_schedule_id = train_schedule.id.clone();
    let train_schedule: TrainSchedule = train_schedule.into();
```

hamz2a replied:
`train_schedule` is `&models::TrainSchedule` and needs to be converted to `schema::TrainSchedule`, so `clone()` is required.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,26 +1136,359 @@ pub(in crate::views) async fn occupancy_blocks(
+}
+
+/// Computes track occupancies for all occurrences of a single paced train.
+fn find_all_track_occupancies_for_paced_train(
```
Not necessary to suffix with `paced_train` because we're already in `paced_train` module.

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/train_schedule.rs:
```diff
@@ -1085,36 +1051,27 @@ fn interpolate_track_occupancy(
+    .collect()
 }
 
 fn find_track_occupancy_for_operational_point(
```
Maybe this function can also be extracted in `track_occupancy_utils.rs`. You might need to use `PacedTrain::into_occurrences()` and make the current function take a `impl Iterator<Item = (TrainId, TrainSchedule)>`.

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,26 +1136,359 @@ pub(in crate::views) async fn occupancy_blocks(
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
+#[schema(as = PacedTrainTrackOccupancyForm )]
```
```suggestion
#[schema(as = PacedTrainTrackOccupancyForm)]
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,26 +1136,359 @@ pub(in crate::views) async fn occupancy_blocks(
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, ToSchema)]
+#[schema(as = PacedTrainTrackOccupancy )]
```
```suggestion
#[schema(as = PacedTrainTrackOccupancy)]
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,26 +1136,359 @@ pub(in crate::views) async fn occupancy_blocks(
+}
+
+#[derive(Debug, Clone)]
+struct TrackOccupancyInput {
```
Maybe we don't need this `struct` at all when using `TrainSchedule::into_occurrences()`?

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,26 +1136,359 @@ pub(in crate::views) async fn occupancy_blocks(
+
+    let path_items = schedules
+        .iter()
+        .flat_map(|ts| ts.path.iter().map(|p| &p.location))
```
minor: 

```suggestion
        .flat_map(|ts| &ts.path)
        .map(|p| &p.location)
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -2103,11 +2389,125 @@ mod tests {
     }
+
+    fn simulation_response() -> core_client::simulation::Response {
+        core_client::simulation::Response::Success(SimulationSuccess {
```
Extract in fixtures file?

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -2103,11 +2389,125 @@ mod tests {
+                blocks: vec![],
+                routes: vec![],
+                track_section_ranges: vec![
+                    TrackRange::new("TA0", 700000, 2000000),
```
Maybe a comment about which track section correspond to which Operational Point.

hamz2a replied:
Done.

woshilapin reviewed (commented):
I believe exploring the `into_occurrence()` is needed, hope it will allow to remove some code you had to do.

hamz2a committed:
editoast: extract track occupancy utils and simplify train_schedule

Signed-off-by: hamz2a <atrari.hamza@gmail.com>

hamz2a committed:
editoast: add track_occupancy endpoint for paced trains

Signed-off-by: hamz2a <atrari.hamza@gmail.com>

woshilapin commented on editoast/src/views/timetable/train_schedule.rs:
```diff
@@ -995,248 +986,66 @@ pub(in crate::views) async fn track_occupancy(
         .zip(train_schedules)
         .flat_map(|((simulation, pathfinding), train_schedule)| {
-            find_track_occupancy_for_operational_point(
+            let train_schedule_schema: TrainSchedule = train_schedule.clone().into();
+            let train_schedule_id = train_schedule.id;
```
```suggestion
            let train_schedule_id = train_schedule.id.clone();
            let train_schedule_schema: TrainSchedule = train_schedule.into();
```

hamz2a replied:
`train_schedule` is `&models::TrainSchedule` and needs to be converted to `schema::TrainSchedule`, so `clone()` is required.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,23 +1138,203 @@ pub(in crate::views) async fn occupancy_blocks(
+        .map(|((train_id, train_schedule), (simulation, pathfinding))| {
+            let occurrence_id: OccurrenceId = match train_id {
+                models::paced_train::TrainId::TrainSchedule(_) => {
+                    return Err(PacedTrainError::UnexpectedTrainScheduleId);
```
This is supposed to be unreachable. But I believe we may not have to care about this case if we modify `PacedTrain::into_occurrences()` to return `OccurrenceId` instead of `TrainId`.

Take a look at #13724

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,23 +1138,203 @@ pub(in crate::views) async fn occupancy_blocks(
+                    )
+                },
+            )
+            .collect::<Vec<_>>())
```
You can use `.collect_vec()` from `itertools`.

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,23 +1138,203 @@ pub(in crate::views) async fn occupancy_blocks(
+            )
+            .collect::<Vec<_>>())
+        })
+        .collect::<Result<Vec<_>, _>>()?
```
You can use `.try_collect()` from `itertools` (but this should disappear with #13724).

woshilapin commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+}
+
+/// Common function to interpolate track occupancy
+pub fn interpolate_track_occupancy(
```
This doesn't need to be public (but you might need to migrate some tests from `train_schedule.rs`).

```suggestion
fn interpolate_track_occupancy(
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+}
+
+/// Common function to compute track occupancy on a specific path item
+pub fn compute_track_occupancy_on_path_item(
```
This doesn't need to be public (but you might need to migrate some tests from `train_schedule.rs`).
```suggestion
fn compute_track_occupancy_on_path_item(
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+
+/// Prepare all data needed for track occupancy computation
+/// This extracts pathfinding data, simulation data, and finds the matching path item
+pub fn prepare_track_occupancy_data<'a>(
```
```suggestion
fn prepare_track_occupancy_data<'a>(
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+
+/// Extract report train from simulation result
+/// Returns None if simulation failed
+pub fn extract_simulation_report_train(simulation: &simulation::Response) -> Option<&ReportTrain> {
```
```suggestion
fn extract_simulation_report_train(simulation: &simulation::Response) -> Option<&ReportTrain> {
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+/// to compute the track occupancy for an invalid train schedule.
+/// It's not a problem because we are just looking for the track occupancy
+/// of the operational point for the valid trains.
+pub fn extract_pathfinding_data(
```
```suggestion
fn extract_pathfinding_data(
```

hamz2a replied:
Done.

woshilapin commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+
+/// Match the id of an operational point with a path item in the train schedule
+/// Returns the path item id if found
+pub fn match_path_item_id_with_operational_point<'a>(
```
```suggestion
fn match_path_item_id_with_operational_point<'a>(
```

hamz2a replied:
Done.

woshilapin reviewed (commented):
Even if the use of `into_occurrences` brings a lot of clarity to the code, I believe there is still some work to do (apart from the tests).

woshilapin committed:
editoast: use OccurrenceId in PacedTrain::into_occurrences()

Since we introduced `OccurrenceId` as a sub-`enum` of `TrainId`,
`PacedTrain::into_occurrences()` should use it now. It brings a lot of
changes in #13671, but I believe for good reasons.

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

leovalais commented on editoast/core_client/src/pathfinding.rs:
```diff
@@ -159,6 +159,18 @@ pub struct TrackRange {
     pub direction: Direction,
 }
 
+#[cfg(feature = "mocking_client")]
+impl TrackRange {
+    pub fn new(track_section: &str, begin: u64, end: u64) -> Self {
+        Self {
+            track_section: track_section.into(),
+            begin,
+            end,
+            direction: Direction::StartToStop,
+        }
+    }
+}
+
```
We already have:

* `schemas::DirectionalTrackRange::new` and a nice DSL to build them concisely
* `From<DirectionalTrackRange> for core_client::TrackRange`

I'd suggest to use these instead of adding a constructor with a fixed parameter.

hamz2a replied:
I find it a bit tricky to do, we've already discussed this with @woshilapin.  
I think it's better to have a simple way to create `TrackRange` like this.

woshilapin replied:
It indeed feels hacky to build another object, then convert it to what you need, it's a weird indirection. That said, we could have the same DSL for `core_client::TrackRange` 🤷 Either way is fine (simple constructor or the same DSL)

hamz2a replied:
as discused, I added the "direction" parameter to the new() constructor.
https://github.com/OpenRailAssociation/osrd/compare/1678fdaaa07ab167d6e64b1f7f5a149d8dd2ce67..9e2497fb69f165ffe2210d184df0859acf3ffd1d#diff-4f1fad5207e287b91a190830fa6fc9ac36a318453fa7e2ee9603922b60b00378R164

leovalais replied:
I haven't seen things this way, I get your point. Let's keep the constructor.

leovalais commented on editoast/src/views/timetable.rs:
```diff
@@ -3,6 +3,7 @@ pub mod paced_train;
 pub(in crate::views) mod similar_trains;
 pub mod simulation;
 pub mod stdcm;
+mod track_occupancy_utils;
```
```suggestion
mod track_occupancy;
```

hamz2a replied:
Done.

leovalais commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+use crate::views::timetable::simulation::SimulationResponseSuccess;
+
+#[derive(Debug, Clone)]
+pub struct TrackOccupancyResult {
```
```suggestion
pub(super) struct TrackOccupancy {
```

In most cases, the `-Result`, `-Data`, etc. suffixes are noise IMO.

hamz2a replied:
Done.

leovalais commented on editoast/src/views/timetable/track_occupancy.rs:
```diff
@@ -0,0 +1,269 @@
+/// Common function to compute track occupancy on a specific path item
+pub fn compute_track_occupancy_on_path_item(
+    train_schedule: &schemas::TrainSchedule,
+    path_item_id: &str,
+    path_item_positions: &[u64],
+    path_projection: &PathProjection,
+    operational_point_track_offsets: &[TrackOffset],
+    report_train: &ReportTrain,
```
Better use `TrackOccupancyData`

hamz2a replied:
I removed `TrackOccupancyData`.

leovalais commented on editoast/src/views/timetable/track_occupancy.rs:
```diff
@@ -0,0 +1,269 @@
+pub fn find_track_occupancy_for_operational_point(
+    operational_point_id: &str,
+    operational_point_track_offsets: &[TrackOffset],
+    path_item_cache: &PathItemCache,
```
Taking in arguent the `PathItemCache` built somewhere else and then querying it is an open door to bugs. You feed the `PathItemCache` some OP descriptors and it queries them and indexes them on its own. However, say you ask it to retrieve an OP using its UIC, you won't be able to query it using its ID. It's a broken API, but that's where we are. So building the `PathItemCache` outside this module, but querying it here **assumes** the cache was built in a very specific way. Therefore this module logic **isn't self-contained at all**. We should either build the cache here, or fix it altogether to give it a reasonable behavior.

hamz2a replied:
The function `find_track_occupancy_for_operational_point` already exists [here](https://github.com/OpenRailAssociation/osrd/blob/dev/editoast/src/views/timetable/train_schedule.rs#L1120). I moved it to `track_occupancy.rs` so it can be used by `paced_trains`. I can refactor it, but in a separate PR.

hamz2a replied:
https://github.com/OpenRailAssociation/osrd/issues/13760

leovalais commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+
+/// Extract report train from simulation result
+/// Returns None if simulation failed
+pub fn extract_simulation_report_train(simulation: &simulation::Response) -> Option<&ReportTrain> {
```
Do we really need a function for a simple match?

hamz2a replied:
Removed.

leovalais commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+
+/// Prepare all data needed for track occupancy computation
+/// This extracts pathfinding data, simulation data, and finds the matching path item
+pub fn prepare_track_occupancy_data<'a>(
```
Feels like a "catch-all" structure (as the `-Data` suffix suggests lacking a more precise description) even though it doesn't "catch all". Here's my suggestion:

* remove the `report_train`
* rename the structure to `PathParameters` (we're already in `mod track_occupancy` and this structure should be private)
* extract the `report_train` directly in `find_track_occupancy_for_operational_point`
* use `PathParameters` in all functions instead of unwrapping it directly in `find_track_occupancy_for_operational_point`

hamz2a replied:
I removed `TrackOccupancyData`.

leovalais commented on editoast/src/views/timetable/track_occupancy_utils.rs:
```diff
@@ -0,0 +1,269 @@
+/// to compute the track occupancy for an invalid train schedule.
+/// It's not a problem because we are just looking for the track occupancy
+/// of the operational point for the valid trains.
+pub fn extract_pathfinding_data(
```
Could be inlined as well. Nice touch on the documentation, however it is invalid as a function documentation.

> We skip it because we don't want
> to compute the track occupancy for an invalid train schedule.

That's not true in the scope of the function, the skipping occurs at call-site. Better IMO just inline the function and place the comment where the skipping actually occurs. Wdyt?

hamz2a replied:
Removed.

leovalais commented on editoast/src/views/timetable/train_schedule.rs:
```diff
@@ -1881,14 +1698,14 @@ pub mod tests {
-        assert_eq!(track, expected_track);
-        assert_eq!(occupancy.train_schedule_id, train_schedule.id);
+        assert_eq!(track_section, expected_track);
+        // assert_eq!(occupancy.train_schedule_id, train_schedule.id);
```
A reason to do so?

hamz2a replied:
I removed the line because we use `schema::TrainSchedule` in `compute_track_occupancy_on_path_item` instead of `models`, so it works with `paced_trains`.

leovalais commented on editoast/src/views/timetable/train_schedule.rs:
```diff
@@ -1782,6 +1594,7 @@ pub mod tests {
         #[case] expected_stop_duration_ms: i64,
     ) {
         // Create test data
+
```
blank line

hamz2a replied:
Done.

leovalais commented on editoast/core_client/src/fixtures.rs:
I'm not a fan of fixtures files personally. They break locality, tie together otherwise independent tests and encourage weird setup patterns (`ProjectStudyScenarioAndSimulationWithTestAppPlusWhatNot`, reuse an existing fixture but mutating it because it's not quite right, etc.). I'd be fine with moving the function upwards in the module tree, but not up to `views/` level, and not with introducing it in the `core_client` even though it's not used for anything in that crate. Lmk what you think.

hamz2a replied:
https://github.com/OpenRailAssociation/osrd/pull/13671#discussion_r2436007453

woshilapin replied:
I'm not particularly one way or the other to be honest. I agree with what you say @leovalais, and at the same time, test data sometimes creates huge noise to understand the real purpose of a test, hence, moving it somewhere else, with a bit of reusability to avoid duplication is a pragmatic solution that breaks the theory you mention. My comment is only a comment in the continuity of what already exists in the repo: since we already have that pattern in the project, let's keep the coherence. Maybe this need some more discussion that may involve more than the scope of the current PR?

hamz2a replied:
as discussed, I created `simulation_empty_response` in `views::timetable`.
https://github.com/OpenRailAssociation/osrd/compare/1678fdaaa07ab167d6e64b1f7f5a149d8dd2ce67..9e2497fb69f165ffe2210d184df0859acf3ffd1d#diff-38ea2dc567249c6105b32817e94bddc17feb13efdef936894de66da4c09327a5R969

leovalais replied:
> test data sometimes creates huge noise to understand the real purpose of a test, hence, moving it somewhere else

100% agreed, but that "somewhere else" doesn't have to be a centralized place for *all* test data. Extracting lengthy setups in local functions somewhere else in the same `mod tests` already serves the clarification purpose. For shared utilities, I'm fine with moving them up the module tree like @hamz2a did. But moving them at crate root should only be for general-purpose tooling (such as `TestApp`) or for fixtures that we literally use everywhere (I guess `small_infra` could fall in this category, but not much else that I can think of).

In retrospect, it should have been obvious that putting all test data in the same place wouldn't be a good idea in the long run.

> My comment is only a comment in the continuity of what already exists in the repo

I just didn't saw your comment before submitting mine 👀. 

> let's keep the coherence

We're already plenty incoherent if you ask me though. A lot of test data are either inlined in their tests, or shared at some local level. There has never been any "rules" on how we want to use `fixtures.rs` afaik.

> Maybe this need some more discussion that may involve more than the scope of the current PR?

Agreed

leovalais commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,28 +1136,201 @@ pub(in crate::views) async fn occupancy_blocks(
+
+    // For each occurrence + simulation result, compute track occupancies
+    let all_occupancies: Vec<(String, TrackOccupancy)> = train_occurrences
+        .iter()
```
```suggestion
        .into_iter()
```

hamz2a replied:
Done.

leovalais commented on editoast/src/views/timetable/paced_train.rs:
```diff
@@ -1130,28 +1136,201 @@ pub(in crate::views) async fn occupancy_blocks(
+    occurrence_id: OccurrenceId,
+    time_begin: DateTime<Utc>,
+    #[schema(value_type = chrono::Duration, example = "PT5M")]
+    duration: PositiveDuration,
```
I know it's only two fields, but organization-wise it would be better imo to put them in their own structure and include it in both train schedule and paced train responses.

hamz2a replied:
I get your point about better organization. I'm not super excited to add a whole new structure just for these two fields, but I'm not against it either. What do you think of this? Could you suggest a good name for the structure?

woshilapin replied:
We settled on `TrackOccupancyTimeWindow` with @hamz2a.

hamz2a replied:
I changed it to `TimeWindow` since we're already in the `track_occupancy` module.
https://github.com/OpenRailAssociation/osrd/compare/1678fdaaa07ab167d6e64b1f7f5a149d8dd2ce67..9e2497fb69f165ffe2210d184df0859acf3ffd1d#diff-e2fe801e211fd49e8942f4b83df0dbeea093d994b88e9d997aafb5585a2df03cR23

leovalais replied:
Nice touch, it's indeed the best name for this object :)

leovalais reviewed (changes requested):
First review, I started before a push force (and lost a few comments in the process, thanks GitHub 😭), I'm not sure how much of it is still relevant. 

I feel like some things are misplaced, and others could be simplified. On a more global note, there's an excessive use of `pub`, many things can be made `pub(in crate::views::timetable)` or even private.

About `OccurrenceId`, it is indeed better than a plain `TrainId`. However I find it annoying to continue building on top of `iter_occurrences` and other `models::PacedTrain` functions depending on `TrainId`/`OccurrenceId`. That will only make the refactoring necessary to move `PacedTrain` in `editoast_models` harder. More about this in https://github.com/OpenRailAssociation/osrd/pull/13037 (and https://github.com/OpenRailAssociation/osrd/pull/13081 that followed). Would it be possible to make alternatives `PacedTrain` functions keeping `OccurrenceId` out of the signature and use them in the new endpoint? After that we'll be able to circle back to `views::train_schedule` as an enhancement. I'd like to limit the amount of new code we introduce we know we'll have to refactor soon.

woshilapin commented on editoast/core_client/src/fixtures.rs:
```diff
@@ -0,0 +1,46 @@
+use crate::simulation::SimulationSuccess;
+use crate::simulation::SpeedLimitProperties;
+
+pub fn simulation_response() -> simulation::Response {
```
Isn't there other places where this fixture can be used, now that it exists? Might be done in another PR if time is short. I'll let you resolve the comment.

hamz2a replied:
https://github.com/OpenRailAssociation/osrd/issues/13764

woshilapin reviewed (approved):
Thanks for all the improvements.

leovalais reviewed (approved):
LGTM, thanks for the changes and the refactoring! I think it's enough for this PR, but as discussed, we'll need at least two follow-up PRs to address:

1. The misplacement of `TrainId` and `OccurrenceId` in `editoast_models`
2. `PathItemCache` not indexing operational points properly

woshilapin commented:
Here are all the follow-up work:
- https://github.com/OpenRailAssociation/osrd/issues/13764
- https://github.com/OpenRailAssociation/osrd/issues/13760
- https://github.com/OpenRailAssociation/osrd/issues/13782

Akctarus commented on front/public/locales/en/errors.json:
```diff
@@ -139,6 +139,7 @@
       "ExceptionNotFound": "Exception '{{exception_key}}' could not be found",
       "InfraNotFound": "Infrastructure '{{infra_id}}' could not be found",
       "NotFound": "Paced train '{{paced_train_id}}' could not be found",
+      "OperationalPointNotFound": "Operational Point '{{operational_point_id}}' could not be found",
```
```suggestion
      "OperationalPointNotFound": "Operational point '{{operational_point_id}}' could not be found",
```
But I saw that “Operational Point” was also written in the same file, so it looks good to me

woshilapin replied:
Fixed in https://github.com/OpenRailAssociation/osrd/compare/1b68ef5c674315c9c5cb37df2ad00bf8d0b37ed5..1e2a057774eca6d29e2e9cb589399e13a9f5645d.

Akctarus reviewed (approved):
LGTM, just left a nit comment

SharglutDev commented on front/src/common/api/generatedEditoastApi.ts:
```diff
@@ -1978,6 +1989,35 @@ export type PostPacedTrainSimulationSummaryApiArg = {
+export type PostPacedTrainTrackOccupancyApiResponse =
+  /** status 200 Track section occupancy periods for paced trains */ {
+    [key: string]: ((
+      | {
+          index: number;
+          paced_train_id: number;
+        }
+      | {
+          exception_key: string;
+          index: number;
+          paced_train_id: number;
+        }
+      | {
+          exception_key: string;
+          paced_train_id: number;
+        }
```
Curious how you feel about it @woshilapin @leovalais but front-end "struggles" dealing with discriminated union like that.

For example, to access the second case, we have to tell typescript `if ('exception_key' in trackOccupancy && 'index' in trackOccupancy)` to make him understand we are on this case and not the 3rd one and can do `trackOccupancy.exception_key` and `trackOccupancy.index` in the same context. 
But this is a bad way to do it because even if these properties don't exist in the type at all, typescript will be happy about it and we may encounter a runtime error if later the api changes and these fields are renamed/removed.

A more efficient way to build the type would be (also removed `paced_train_id` from this part, it is common to all cases so can be moved with `duration` and `time_begin`) : 
```
| {
          index: number;
          type: 'indexed_occurrence';
        }
      | {
          exception_key: string;
          index: number;
          type: 'indexed_exception';
        }
      | {
          exception_key: string;
          type: 'added_exception';
        }
```
This way, as `type` is common in all cases, we can easily access each case by just doing `if (type === xxx)`, typescript will be happy, if the api changes later we can spot typescript error very easily as well.

Many editoast endpoints are like that and it would be great to have an enhancement to update them with this architecture if you agree and if openapi allows it.

woshilapin replied:
Pushed in https://github.com/OpenRailAssociation/osrd/pull/13671/commits/1b68ef5c674315c9c5cb37df2ad00bf8d0b37ed5.

SharglutDev replied:
Nice, thanks for the change !

One last thing that went probably unnoticed in my large comment, `paced_train_id` could be moved alongside `duration` and `time_begin` since it is common in all cases.

woshilapin replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/13671/commits/58e11b362b1922a1b6725704fb02d85fa63b69cf

woshilapin committed:
editoast: make tagged `OccurrenceId` for the front

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
editoast: factorize 'paced_train_id' in 'TrainId'

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

SharglutDev commented on front/src/common/api/generatedEditoastApi.ts:
```diff
@@ -1978,6 +1989,37 @@ export type PostPacedTrainSimulationSummaryApiArg = {
+          exception_key: string;
+          type: 'CreatedException';
+        }
+    ) & {
+      duration: string;
+      time_begin: string;
+    } & {
+      paced_train_id: number;
+    })[];
```
This is weird 🤔  Why are they not together ? Is it openapi ?

woshilapin replied:
Yes, it's expected because of the generation tools. On `editoast` side, the `struct` looks like this

```rust
struct TimeWindow {
    time_begin: DateTime<Utc>,
    duration: PositiveDuration,
}

struct TrackOccupancy {
    paced_train_id: i64,
    #[serde(flatten)]
    #[schema(inline)]
    occurrence_id: OccurrenceId,
    #[serde(flatten)]
    #[schema(inline)]
    time_window: track_occupancy::TimeWindow,
}
```

So I'm guessing the auto-generation tools are considering the `paced_train_id` in a different group than the `TimeWindow` which contains the 2 other fields.

SharglutDev replied:
Ok I understand

SharglutDev reviewed (approved):
Lgtm, thanks for the changes ! (if they can stay 🙂)
